service/qos/service_level_controller.cc

```diff
@@ -687,6 +688,66 @@ future<> service_level_controller::unregister_subscriber(qos_configuration_chang
     return _subscribers.remove(subscriber);
 }
 
+static sstring describe_service_level(std::string_view sl_name, const service_level_options& sl_opts) {
+    using slo = service_level_options;
+
+    utils::small_vector<sstring, 2> opts{};
+    
+    const sstring sl_name_formatted = cql3::util::maybe_quote(sl_name);
+
+    if (auto maybe_timeout = std::get_if<lowres_clock::duration>(&sl_opts.timeout)) {
+        // According to the documentation, `TIMEOUT` has to be expressed in milliseconds
+        // or seconds. It is therefore safe to use milliseconds here.
+        const auto timeout = std::chrono::duration_cast<std::chrono::milliseconds>(*maybe_timeout);
+        opts.push_back(seastar::format("TIMEOUT = {}", timeout));
+    }
+
+    switch (sl_opts.workload) {
+        case slo::workload_type::batch:
+            opts.push_back("WORKLOAD_TYPE = 'batch'");
+            break;
+        case slo::workload_type::interactive:
+            opts.push_back("WORKLOAD_TYPE = 'interactive'");
+            break;
+        case slo::workload_type::unspecified:
+            break;
+        case slo::workload_type::delete_marker:
+            on_internal_error(sl_logger, "Precondition has been violated");
+    }
+
+    if (opts.size() == 0) {
+        return seastar::format("CREATE SERVICE_LEVEL {};", sl_name_formatted);
+    }
+
+    return seastar::format("CREATE SERVICE_LEVEL {} WITH {};", sl_name_formatted, fmt::join(opts, " AND "));
+}
+
+future<std::vector<cql3::description>> service_level_controller::describe_service_levels() const {
+    // We rely on the assumption that `_service_levels_db` (so the cache of
+    // the `service_level_controller`) is always up-to-date.
```

Comment:
`_service_levels_db` functions as a cache for `service_level_controller`. It keeps track of the state of service levels. However, since it's just a cache, it might not keep the current state. As @Jadw1 explained in [his comment](https://github.com/scylladb/scylladb/pull/20168#discussion_r1723856118), the cache is always up-to-date when Raft is enabled. When implementing it, however, I forgot about the fact that we rely on using Raft, so we'll probably need a branch here